examples/audio/transformer_asr.py

```diff
@@ -0,0 +1,391 @@
+"""
+Title: Automatic Speech Recognition with Transformer
+Author: [Apoorv Nandan](https://twitter.com/NandanApoorv)
+Date created: 2021/01/13
+Last modified: 2021/01/13
+Description: Implement a complete transformer and train it on sequence to sequence ASR.
+"""
+"""
+## Introduction
+Automatic speech recognition is the task of predicting text from an audio segment.
+One of the popular ways to solve this task, is by treating it as a sequence to
+sequence problem. The audio can be represented as a sequence of feature vectors
+and the text can be considered as a sequence of characters, words, or subword tokens.
+ASR models typically require huge datasets and take a lot of time to train.
+For this demonstration, we will use a simple dataset to reduce the training time and
+display some results. Our model will be a complete transformer (both encoder and decoder)
+as proposed in the paper, "Attention is All You Need".
+Our data consists of audio segments with a person speaking out one of the ten digits
+(0-9). We convert the digits as a sequence of characters (e.g. 9 -> n,i,n,e) to form
+a small dataset. The method shown below, however, can be applied to a real world
+speech dataset, to learn the mapping from audio to a sequence of characters.
+**References:**
+- [Attention is All You Need](https://papers.nips.cc/paper/2017/file/3f5ee243547dee91fbd053c1c4a845aa-Paper.pdf)
+- [Free Spoken Digit Dataset](https://github.com/Jakobovski/free-spoken-digit-dataset)
+"""
+import os
+import random
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import layers
+
+tf.__version__
+"""
+## Transformer Input Layer
+This layer computes the sum of position embeddings and feature embeddings to feed to
+the transformer layers.
+"""
+
+
+class TransformerInput(layers.Layer):
+    def __init__(
+        self, input_type="tokens", nvocab=1000, nhid=64, nff=128, maxlen=100,
+    ):
+        super().__init__()
+        self.input_type = input_type
+        if input_type == "tokens":
+            self.emb = tf.keras.layers.Embedding(nvocab, nhid)
+        elif input_type == "feats":
+            self.emb = tf.keras.layers.Dense(nhid)
```

Comment:
You might want to handle the 3rd case with an error or otherwise have some form of value validation for "input_type"